azure-kusto-data/source/tokenProvider.ts

```diff
@@ -0,0 +1,293 @@
+                let resourceUri = this.cloudInfo.KustoServiceResourceId;
+                if (this.cloudInfo.LoginMfaRequired) {
+                    resourceUri = resourceUri.replace(".kusto.", ".kustomfa.")
+                }
+                this.scopes = [resourceUri + "/.default"]
+                this.initClient();
+            }
+            token = await this.acquireMsalToken();
+            this.initialized = true;
+        }
+        else {
+            const accounts = await this.msalClient?.getTokenCache()?.getAllAccounts();
+            if (accounts?.length > 0) {
+                token = await this.msalClient.acquireTokenSilent({ scopes: this.scopes, account: accounts[0] });
+            }
+            if (token == null) {
+                token = await this.acquireMsalToken();
+            }
+        }
+
+        if (token != null) {
+            return { tokenType: token.tokenType, accessToken: token.accessToken }
+        }
+        throw new Error("Failed to get token from msal");
+    }
+}
+
+/**
+ * Acquire a token from MSAL with username and password
+ */
+export class UserPassTokenProvider extends MsalTokenProvider {
+    name = "UserPassTokenProvider";
+    userName: string;
+    password: string;
+
+    constructor(kustoUri: string, userName: string, password: string, authorityId?: string) {
+        super(kustoUri, authorityId);
+        this.userName = userName;
+        this.password = password;
+    }
+
+    initClient(): void {
+        const clientConfig = {
+            auth: {
+                clientId: this.cloudInfo.KustoClientAppId,
+                authority: CloudSettings.getAuthorityUri(this.cloudInfo, this.authorityId),
+            }
+        };
+        this.msalClient = new PublicClientApplication(clientConfig);
+    }
+
+    acquireMsalToken(): Promise<AuthenticationResult | null> {
+        return (this.msalClient as PublicClientApplication).acquireTokenByUsernamePassword({ scopes: this.scopes, username: this.userName, password: this.password });
+    }
+}
+
+/**
+ * Acquire a token from MSAL with Device Login flow
+ */
+export class DeviceLoginTokenProvider extends MsalTokenProvider {
+    name = "DeviceLoginTokenProvider";
+    deviceCodeCallback: (response: DeviceCodeResponse) => void;
+
+    constructor(kustoUri: string, deviceCodeCallback: (response: DeviceCodeResponse) => void, authorityId?: string) {
+        super(kustoUri, authorityId);
+        this.deviceCodeCallback = deviceCodeCallback;
+    }
+
+    initClient(): void {
+        const clientConfig = {
+            auth: {
+                clientId: this.cloudInfo.KustoClientAppId,
+                authority: CloudSettings.getAuthorityUri(this.cloudInfo, this.authorityId),
+            },
+        };
+        this.msalClient = new PublicClientApplication(clientConfig);
+    }
+
+    acquireMsalToken(): Promise<AuthenticationResult | null> {
```

Comment:
Have you tested this?
It requires some user interaction and in python it prints a URL and a code to provide

Reply:
I'm basically asking how does this work in node

Reply:
Tested